UM/FlameProfiler.py

```diff
@@ -0,0 +1,187 @@
+def stopRecordingProfileData():
+    global stop_record_profile_requested
+    stop_record_profile_requested = True
+
+def _fillInProfileSpaces(start_time, end_time, profile_call_list):
+    result = []
+    time_counter = start_time
+    for profile_call in profile_call_list:
+        if secondsToMS(profile_call.getStartTime()) != secondsToMS(time_counter):
+            result.append(_ProfileCallNode("", 0, time_counter, profile_call.getStartTime(), []))
+        result.append(profile_call)
+        time_counter = profile_call.getEndTime()
+
+    if secondsToMS(time_counter) != secondsToMS(end_time):
+        result.append(_ProfileCallNode("", 0, time_counter, end_time, []))
+
+    return result
+
+def secondsToMS(value):
+    return math.floor(value *1000)
+
+##  Profile a block of code.
+#
+#   Use this context manager to wrap and profile a block of code.
+#   \param name \type{str} The name to use to identify this code in the profile report.
+@contextmanager
+def profileCall(name):
+    start_time = time.time()
+    child_accu_stack.append([])
+    yield
+    end_time = time.time()
+    call_stat = _ProfileCallNode(name, 0, start_time, end_time,
+                                 _fillInProfileSpaces(start_time, end_time, child_accu_stack.pop()))
+    child_accu_stack[-1].append(call_stat)
+
+##  Return whether we are recording profiling information.
+#
+#   \return \type{bool} True if we are recording.
+def isRecordingProfile():
+    global record_profile
+    return record_profile
+
+
+def updateProfileConfig():
+    global child_accu_stack
+    global record_profile
+
+    # We can only update the active profiling config when we are not deeply nested inside profiled calls.
+    if len(child_accu_stack) <= 1:
+        global clear_profile_requested
+        if clear_profile_requested:
+            clear_profile_requested = False
+            child_accu_stack = [[]]
+
+        global record_profile_requested
+        if record_profile_requested:
+            record_profile_requested = False
+            record_profile = True
+            Logger.log('d', 'Starting record record_profile_requested')
+
+        global stop_record_profile_requested
+        if stop_record_profile_requested:
+            stop_record_profile_requested = False
+            record_profile = False
+            Logger.log('d', 'Stopping record stop_record_profile_requested')
+
+##  Decorator which can be manually applied to methods to record profiling information.
+#
+#   \type{Callable}
+def profile(function):
+    def runIt(*args, ** kwargs):
+        if isRecordingProfile():
+            with profileCall(function.__qualname__):
+                return function(*args, ** kwargs)
+        else:
+            return function(*args, **kwargs)
+    return runIt
+
+##  Drop in replacement for PyQt5's pyqtSlot decorator which records profiling information.
+#
+#   See the PyQt5 documentation for information about pyqtSlot.
+def pyqtSlot(*args, **kwargs):

```

Comment:
I would like to have a switch that can disable this behaviour completely, preferably with an environment variable like I have done for Signal tracing. That way we can enable this only when we actually need it.